`Exam mode`: Allow instructors to verify attendance check for students

pellumbbaboci opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

- It would be important that the instructor immediately identify if a student has started the exam without being in the lecture hall, as this is not allowed and would be considered as cheating
- When removing all students from an exam, the images cannot be deleted, there is an error message on the server FileService.manageFilesForUpdatedFilePath: 
`Could not delete old file '/api/files/exam-user/39411/ExamUserImage_2023-02-12T11-35-22-366_40ba6781.png' due to exception Unknown Filepath: /api/files/exam-user/39411/ExamUserImage_2023-02-12T11-35-22-366_40ba6781.png`

- It would be good if the web client can cache the downloaded images for the exam users, so that they do not need to be downloaded multiple times when navigating through the exam users on the page

### Description
<!-- Describe your changes in detail -->

In this PR the following are implemented:
- New endpoint to identify students who did not sign 
``` GET courses/{courseId}/exams/{examId}/verify-exam-users : Retrieves a list of students who started the exam but did not sign ```

- Cahce images of exam users
- Delete old files (exam user images)


### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

Prerequisites:
- 1 Instructor
- 1 Course with Exam
- Multiple students registered to the Course

1. Log in to Artemis
2. Navigate to Course Management and go under Exams as instructor
3. You should prepare the exam and generate the student exams.
4. Now login as student and navigate to the exam.
5. Start the exam as student (you can start the exam for all students)
6. Now login as instructor and  **After exam starts** check the students by letting them sign (This step can be done through Ipad app or through postman see below)
7. **After exam starts** navigate to students as Instructor
8. You should see a new button called **Verify Attendance** , click it
9. You should see all students that started the exam but did not sign. 

Note:In Order for students to sign, you need to have IPad and the exam-attendance-checker app downloaded. But if you want you can change/update exam-user information by making a request to the following endpoint (use postman): 

``` 
/**
     * POST /courses/:courseId/exams/:examId/exam-users : Update the exam user with the given DTO info
     *
     * @param examUserDTO   the dto containing exam user info
     * @param signatureFile the signature of the student
     * @param courseId      the id of the course
     * @param examId        the id of the exam
     * @return saved examUser ResponseEntity with status 200 (OK) or with status 404 (Not Found)
     */
    @PostMapping("courses/{courseId}/exams/{examId}/exam-users")
    @PreAuthorize("hasRole('INSTRUCTOR')")
    public ResponseEntity<ExamUser> updateExamUser(@RequestPart ExamUserDTO examUserDTO, @RequestPart(value = "file", required = false) MultipartFile signatureFile,
            @PathVariable Long courseId, @PathVariable Long examId) {

....................
}
 ```
Here you can see an postman request example: 

![exampleOfPostman](https://user-images.githubusercontent.com/26312757/219153178-b1f33a07-8f81-4874-9185-72112ddc5bcc.png)
examUserDTO object is as follow:

```
{
 "login": "student",
"didCheckImage":true,
"didCheckLogin":true,
"didCheckName":true,
"didCheckRegistrationNumber":true
}
```
Note: login field is important as its used as identifier for students

#### Exam Mode Testing
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
10. Make sure that the UI of the programming exercise in the exam mode stays unchanged. You can use the [exam mode documentation](https://docs.artemis.cit.tum.de/user/exam_mode/) as reference.
11. ...

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [ ] Test

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- Note: You may use the table below or copy the file coverage from the Codecov bot's comment. -->
<!--
| Class/File | Branch | Line |
|------------|-------:|-----:|
| ExerciseService.java | 85% | 77% |
| programming-exercise.component.ts | 13% | 95% |
-->

**Server**
| Class/File | Branch | Line |
|------------|-------:|-----:|
| ExamUserAttendanceCheckDTO.java | n/a | 100% |
| FileService.java | 85% | 87% |
| ExamUserResource.java | 75% | 91% |
| FileResource.java | 60% | 77% |

**Client**
| Class/File | Branch | Line |
|------------|-------:|-----:|
| exam-management.service.ts | 76.46%| 95.69% |
| exam-students.component.ts | 69.56%| 86.59% |
| students-upload-images-dialog.component.ts | 48.38%| 84.44% |
| exam-students-attendance-check.component.ts | 74.35%| 98.38% |
| exam-users-attendance-check.model.ts | 100%| 0% |



### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/201
7/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

New button to access the verify attendance page:
![verifyButton](https://user-images.githubusercontent.com/26312757/219146916-216e411f-3483-40b3-954e-8338e23b86a2.png)


Shows students that were not checked:
![notCheckedUsers](https://user-images.githubusercontent.com/26312757/219147123-80de912b-c988-4c83-bc96-0a74a0a899f5.png)

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExamUserResource.java:
```diff
@@ -107,4 +105,19 @@ public ResponseEntity<ExamUsersNotFoundDTO> saveUsersImages(@PathVariable Long c
+     * @param examId   the id of the exam
+     * @return list of students who did not sign ResponseEntity with status 200 (OK)
+     */
+    @GetMapping("courses/{courseId}/exams/{examId}/exam-users-verify")
```
```suggestion
    @GetMapping("courses/{courseId}/exams/{examId}/verify-exam-users")
```

Strohgelaender commented on src/main/webapp/app/exam/manage/exam-management.service.ts:
```diff
@@ -231,6 +232,16 @@ export class ExamManagementService {
+     * @param examId
+     * @return matriculation number of students that were not found in the system
+     */
+    verifyExamUserAttendanceCheck(courseId: number, examId: number): Observable<HttpResponse<ExamUserAttendanceCheck[]>> {
+        return this.http.get<ExamUserAttendanceCheck[]>(`${this.resourceUrl}/${courseId}/exams/${examId}/exam-users-verify`, { observe: 'response' });
```
```suggestion
    verifyExamUserAttendance(courseId: number, examId: number): Observable<HttpResponse<ExamUserAttendanceCheck[]>> {
        return this.http.get<ExamUserAttendanceCheck[]>(`${this.resourceUrl}/${courseId}/exams/${examId}/verify-exam-users`, { observe: 'response' });
```

Strohgelaender commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.ts:
```diff
@@ -0,0 +1,108 @@
+        private accountService: AccountService,
+        private sortService: SortService,
+    ) {
+        this.predicate = 'id';
```
Move this from the constructor to the declaration above.

Strohgelaender commented on src/main/webapp/i18n/en/exam.json:
```diff
@@ -454,6 +467,10 @@
+                "verifyAttendanceTooltip": "Verify the attendance check of the students",
+                "verifyAttendanceBtn": "Verify Attendance",
+                "examNotStarted": "The exam has not started yet",
+                "verifyChecks": "Verify Attendance Checks",
```
```suggestion
                "verifyChecks": "Verify Attendance",
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/ExamUserRepository.java:
```diff
@@ -20,5 +21,22 @@ public interface ExamUserRepository extends JpaRepository<ExamUser, Long> {
+    List<ExamUser> findAllByExamId(@Param("examId") long examId);
+
+    @Query("""
+            SELECT new de.tum.in.www1.artemis.web.rest.dto.ExamUserAttendanceCheckDTO(examUser.id, examUser.studentImagePath, examUser.user.login,
```
Should we do the object creation directly here or move this into a service?

pellumbbaboci replied:
I think we need to do it here as we need to mix information from `exam_user` and `students_exam` tables. Do you know how to create this object from a service?

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/ExamUserRepository.java:
```diff
@@ -20,5 +21,22 @@ public interface ExamUserRepository extends JpaRepository<ExamUser, Long> {
     ExamUser findByExamIdAndUserId(@Param("examId") long examId, @Param("userId") long userId);
 
-    List<ExamUser> findAllByExamId(long examId);
+    List<ExamUser> findAllByExamId(@Param("examId") long examId);
```
Seems unneccary to me, the `@Param` is only used in `@Query`.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/ExamUserRepository.java:
```diff
@@ -20,5 +21,22 @@ public interface ExamUserRepository extends JpaRepository<ExamUser, Long> {
+            SELECT new de.tum.in.www1.artemis.web.rest.dto.ExamUserAttendanceCheckDTO(examUser.id, examUser.studentImagePath, examUser.user.login,
+            examUser.user.registrationNumber, examUser.signingImagePath,
+            studentExams.started, studentExams.submitted)
+            FROM ExamUser examUser
+            LEFT JOIN examUser.exam exam
+            LEFT JOIN exam.studentExams studentExams ON studentExams.user.id = examUser.user.id
+            WHERE (exam.id = :#{#examId})
```
```suggestion
            FROM ExamUser examUser
                LEFT JOIN examUser.exam exam
                LEFT JOIN exam.studentExams studentExams ON studentExams.user.id = examUser.user.id
            WHERE (exam.id = :examId)
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExamUserResource.java:
```diff
@@ -107,4 +105,19 @@ public ResponseEntity<ExamUsersNotFoundDTO> saveUsersImages(@PathVariable Long c
     }
+
+    /**
+     * POST courses/{courseId}/exams/{examId}/exam-users-verify : Retrieves a list of students who started the exam but did not sign
```
```suggestion
     * POST courses/{courseId}/exams/{examId}/verify-exam-users : Retrieves a list of students who started the exam but did not sign
```

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/repository/ExamUserRepository.java:
```diff
@@ -20,5 +21,22 @@ public interface ExamUserRepository extends JpaRepository<ExamUser, Long> {
+                    OR examUser.didCheckRegistrationNumber = false
+                    OR examUser.didCheckName = false)
+            """)
+    List<ExamUserAttendanceCheckDTO> findAllExamUsersWhoDidNotSign(@Param("examId") long examId);
```
Why a `List`? I think a `Set` makes more sense here, there is no order.

pellumbbaboci replied:
What do you mean by order? Is it order-by clause? And i don't understand why `Set` would make more sense?

Strohgelaender replied:
A `Set` is an unordered collection while a `List` has a order. But since you don't define a `ORDER BY` clause, I'd consider the order as not relevate and would use a `Set` here.

pellumbbaboci replied:
Changed it to `Set`, :+1:

Strohgelaender commented on src/main/webapp/app/exam/manage/students/exam-students.component.ts:
```diff
@@ -95,6 +98,7 @@ export class ExamStudentsComponent implements OnInit, OnDestroy {
         this.isLoading = true;
         this.examManagementService.find(this.courseId, this.exam.id!, true).subscribe((examResponse: HttpResponse<Exam>) => {
             this.exam = examResponse.body!;
```
This block is the same as above, can you move this into a method?

Strohgelaender commented on src/main/webapp/app/exam/manage/students/upload-images/students-upload-images-dialog.component.ts:
```diff
@@ -61,7 +60,7 @@ export class StudentsUploadImagesDialogComponent implements OnDestroy {
     private resetDialog() {
         this.isParsing = false;
-        this.notFoundUsers = undefined;
+        this.notFoundUsers = {} as NotFoundExamUserType;
```
This does not seem correct to me. Accrding to the types, `numberOfUsersNotFound` cannot be undefined, so `this.notFoundUsers .numberOfUsersNotFound` should also not be undefined. But with this assignement this is possible.

pellumbbaboci replied:
Yes you are right. I reverted the change back anyways so this field can be undefined.

Strohgelaender commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.ts:
```diff
@@ -0,0 +1,106 @@
+import { Component, OnDestroy, OnInit, ViewChild, ViewEncapsulation } from '@angular/core';
```
unused imports.

Strohgelaender commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.ts:
```diff
@@ -0,0 +1,106 @@
+    searchFailed = false;
+    searchNoResults = false;
+    isTransitioning = false;
+    rowClass: string | undefined = undefined;
```
```suggestion
    rowClass?: string;
```

Strohgelaender commented on src/main/webapp/i18n/de/exam.json:
```diff
@@ -439,12 +439,25 @@
+                "searchStudents": "Studenten nach Namen oder Login suchen",
                 "searchNoResults": "Keine Studierenden gefunden",
+                "signingImagePath": "Unterschrift",
+                "checks": "Schecks",
```
```suggestion
                "checks": "Überprüfungen",
```

Strohgelaender commented on src/main/webapp/i18n/de/exam.json:
```diff
@@ -453,6 +466,10 @@
                 "registrationNumber": "Matrikelnummer",
                 "registerAllFromCourse": "Registriere alle Studierenden des Kurses",
                 "registerAllFromCourseTooltip": "Registriere alle zu diesem Kurs angemeldeten Studierende",
+                "verifyAttendanceTooltip": "Überprüfung der Anwesenheitskontrolle der Studenten",
```
```suggestion
                "verifyAttendanceTooltip": "Überprüfung der Anwesenheitskontrolle der Studierenden",
```

Strohgelaender commented on src/main/webapp/i18n/de/exam.json:
```diff
@@ -439,12 +439,25 @@
                 "usersPerPage": "{{number}} Studierende pro Seite",
                 "showAllUsers": "Alle Studierenden anzeigen",
                 "searchForUsers": "Studierende für diese Klausur anmelden durch Suchen nach Login oder Namen",
+                "searchStudents": "Studenten nach Namen oder Login suchen",
```
```suggestion
                "searchStudents": "Studierende nach Namen oder Login suchen",
```

Strohgelaender commented on src/main/webapp/i18n/de/exam.json:
```diff
@@ -439,12 +439,25 @@
+                "registrationNumberChecked": "Matrikelnummer wurde erfolgreich geprüft",
+                "registrationNumberNotChecked": "Matrikelnummer wurde nicht überprüft",
                 "registeredStudents": "Angemeldete Studierende",
+                "notCheckedStudents": "Nicht überprüfte Studenten",
```
```suggestion
                "notCheckedStudents": "Nicht überprüfte Studierende",
```

Intektor reviewed (approved):
I checked the *Java* code and apart from potential beauty optimization (which we do not have time for) the code looks quite fine and I could not find anything I dislike.

sven0311 reviewed (dismissed):
Tested Manually and worked perfectly

mertyldrr reviewed (approved):
tested in the testing session on TS10 and works as expected

ge65cer commented on src/main/webapp/app/entities/exam-users-attendance-check.model.ts:
```diff
@@ -0,0 +1,11 @@
+import { BaseEntity } from 'app/shared/model/base-entity';
+
+export class ExamUserAttendanceCheck implements BaseEntity {
```
Is this a DTO? Then it should be named so.

pellumbbaboci replied:
Yes its a DTO, i renamed it. :+1:

ge65cer commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.html:
```diff
@@ -0,0 +1,85 @@
+        <div *ngIf="!hasExamStarted">
+            <h3><span class="text-white badge bg-danger" [jhiTranslate]="'artemisApp.examManagement.examStudents.examNotStarted'">Exam have not started yet!</span></h3>
+        </div>
+        <div *ngIf="hasExamStarted">
```
Please use "else". https://angular.io/api/common/NgIf

ge65cer commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.html:
```diff
@@ -0,0 +1,85 @@
+    </div>
+    <div class="row"></div>
+    <br />
+    <div class="table-responsive" id="exams-table" *ngIf="exam">
```
You check for `exam` here but also use `exam.title` above.

pellumbbaboci replied:
This page is not accessible if there is not exam so i removed the if check because it does not make sense.

ge65cer commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.html:
```diff
@@ -0,0 +1,85 @@
+                        <img *ngIf="examUser?.signingImagePath" class="img-fluid" [width]="60" [src]="SERVER_API_URL + examUser.signingImagePath" />
+                        <h3 *ngIf="!examUser?.signingImagePath">
+                            <span class="m-2">
+                                <fa-icon [ngStyle]="{ color: 'red' }" [icon]="faXmark"></fa-icon>
```
This doesn't need to be a computed property if it is static. Use a Boostrap color class instead, please.

ge65cer commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.html:
```diff
@@ -0,0 +1,85 @@
+                        </span>
+                    </td>
+                    <td class="d-md-table-cell">
+                        <img *ngIf="examUser?.signingImagePath" class="img-fluid" [width]="60" [src]="SERVER_API_URL + examUser.signingImagePath" />
```
`width` doesn't need to be a computed property if it is static.

ge65cer reviewed (changes requested):
Please also address all open change requests from #6137, if not done already.

ge65cer commented on src/main/webapp/app/exam/manage/students/verify-attendance-check/exam-students-attendance-check.component.html:
```diff
@@ -0,0 +1,85 @@
+                    </td>
+                    <td class="d-md-table-cell">
+                        <img *ngIf="examUser?.signingImagePath" class="img-fluid" [width]="60" [src]="SERVER_API_URL + examUser.signingImagePath" />
+                        <h3 *ngIf="!examUser?.signingImagePath">
```
Optional: Please use an "else" block instead.

pellumbbaboci commented:
> Please also address all open change requests from #6137, if not done already.

I have addressed your suggestions but i still have to address some other suggestion and add a db migration. This will be done in another PR from this `chore/exam-code-quality-improvements`

TimOrtel reviewed (approved):
*Java* code looks fine to me. I do not think we should focus too much on the beauty of this code as we need to merge this soon.

bassner reviewed (approved):
Checked the code and it looks good, given the priority of the feature. It will work fine. We can follow up with some code polish PR later on :)

JohannesStoehr reviewed (approved):
Code looks good to me

tobias-lippert reviewed (approved):
Code LGTM

sven0311 reviewed (approved):
Manually Reviewed with the iPad app and works like a charm

danielbode reviewed (approved):
Tested on ts2 using postman. Works as expected. No bugs found. Nice feature!

Just a quick note if you also want to test it with postman and don't have much experience with it. You also need to add a valid JWT to the header. Otherwise, you don't have permission to POST to the API endpoint. You can get a valid token e.g. by logging in as an instructor in your browser, inspecting the network traffic in the dev tools, and copying it from header -> request -> cookie.

pellumbbaboci commented:
> Just a quick note if you also want to test it with postman and don't have much experience with it. You also need to add a valid JWT to the header. Otherwise, you don't have permission to POST to the API endpoint. You can get a valid token e.g. by logging in as an instructor in your browser, inspecting the network traffic in the dev tools, and copying it from header -> request -> cookie.

Thanks for the hint!
